### PR TITLE
Merge Switch and Button Drivers

### PR DESCRIPTION
**Pull request by nickolas-deboom**
# Type of Change

- [ ] WWST Certification Request
     - If this is your first time contributing code:
          - [ ] I have reviewed the README.md file
          - [ ] I have reviewed the CODE_OF_CONDUCT.md file
          - [ ] I have signed the CLA
     - [ ] I plan on entering a WWST Certification Request or have entered a request through the WWST Certification console at developer.smartthings.com
- [ ] Bug fix
- [ ] New feature
- [X] Refactor

# Checklist

- [X] I have performed a self-review of my code
- [X] I have commented my code in hard-to-understand areas
- [X] I have verified my changes by testing with a device or have communicated a plan for testing
- [x] I am adding new behavior, such as adding a sub-driver, and have added and run new unit tests to cover the new behavior

# Description of Change

This PR addresses [CHAD-13270](https://smartthings.atlassian.net/browse/CHAD-13270) to integrate the functionality from matter-button into the matter-switch driver, so that we can deprecate the matter-button driver.

Note that these changes were originally located in [PR 1423](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/pull/1423), but that PR was split into multiple PRs in order to simplify the amount of changes. This PR only includes support for pure-button and pure-switch devices. Once more discussion has been had on how combination button-switch devices should be configured, [PR 1596](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/pull/1596) will be updated to include support for these combination devices.

# Summary of Completed Tests

Testing was performed on several switch and button devices as documented here: https://smartthings.atlassian.net/wiki/spaces/CHIP/pages/3340304693/Matter-Switch+Matter-Button+Integration+Testing+Status

[CHAD-13270]: https://smartthings.atlassian.net/browse/CHAD-13270?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 1]
```diff
@@ -75,6 +84,47 @@ local child_device_profile_overrides = {
...
+-- TODO: add an is_feature_supported_for_endpoint function to matter.device that takes an endpoint
+local EMULATE_HELD = "__emulate_held" -- for non-MSR (MomentarySwitchRelease) devices we can emulate this on the software side
+local SUPPORTS_MULTI_PRESS = "__multi_button" -- for MSM devices (MomentarySwitchMultiPress), create an event on receipt of MultiPressComplete
+local INITIAL_PRESS_ONLY = "__initial_press_only" -- for devices that support MS (MomentarySwitch), but not MSR (MomentarySwitchRelease)
+
+local HUE_MANUFACTURER_ID = 0x100B
+
+--helper function to create list of multi press values
+local function create_multi_list(size, supportsHeld)
+  local list = {"pushed", "double"}
```
Comment (hcarter-775) [inline drivers/SmartThings/matter-switch/src/init.lua:111]: can we get a more clear name? I think in general helper functions should support as high a level of clarity as anything else

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 2]
```diff
@@ -75,6 +84,47 @@ local child_device_profile_overrides = {
...
+  if supportsHeld then table.insert(list, "held") end
+  for i=3, size do
```
Comment (hcarter-775) [inline drivers/SmartThings/matter-switch/src/init.lua:113]: where does 3 come from? Can we get a comment or initialize a clearly defined variable name with a 3

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 3]
```diff
@@ -103,9 +170,19 @@ end
...
 --- BRIDGED_NODE_DEVICE_TYPE on bridged devices
 local function find_default_endpoint(device, component)
-  local eps = device:get_endpoints(clusters.OnOff.ID)
-  table.sort(eps)
-  for _, v in ipairs(eps) do
+  local switch_eps = device:get_endpoints(clusters.OnOff.ID)
+  local button_eps = device:get_endpoints(clusters.Switch.ID, {feature_bitmap=clusters.Switch.types.Feature.MOMENTARY_SWITCH})
+  local all_eps = {}
+
+  for _,v in ipairs(switch_eps) do
```
Comment (hcarter-775) [inline drivers/SmartThings/matter-switch/src/init.lua:177]: I think we should have a clearer variable name than v, for everything below as well

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 4]
```diff
@@ -103,9 +170,19 @@ end
 --- and supports the OnOff cluster. This is done to bypass the
 --- BRIDGED_NODE_DEVICE_TYPE on bridged devices
 local function find_default_endpoint(device, component)
-  local eps = device:get_endpoints(clusters.OnOff.ID)
-  table.sort(eps)
-  for _, v in ipairs(eps) do
+  local switch_eps = device:get_endpoints(clusters.OnOff.ID)
+  local button_eps = device:get_endpoints(clusters.Switch.ID, {feature_bitmap=clusters.Switch.types.Feature.MOMENTARY_SWITCH})
+  local all_eps = {}
```
Comment (hcarter-775) [inline drivers/SmartThings/matter-switch/src/init.lua:175]: can we call this endpoint_list or something like that? all_eps is vague in my opinion

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 5]
```diff
@@ -144,92 +221,200 @@ local function assign_child_profile(device, child_ep)
...
-            type = "EDGE_CHILD",
-            label = name,
-            profile = child_profile,
-            parent_device_id = device.id,
-            parent_assigned_child_key = string.format("%d", ep),
-            vendor_provided_label = name
-          }
-        )
+  local button_eps = device:get_endpoints(clusters.Switch.ID, {feature_bitmap=clusters.Switch.types.Feature.MOMENTARY_SWITCH})
+  local all_eps = {}
```
Comment (hcarter-775) [inline drivers/SmartThings/matter-switch/src/init.lua:268]: can we re-name this one as above

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 6]
```diff
@@ -144,92 +221,200 @@ local function assign_child_profile(device, child_ep)
...
+
+  local new_profile = nil
+
+  local component_map = {}
+  local current_component_number = 2
+  local component_map_used = false
+  local parent_child_device = false
+
+  if #switch_eps > 0 or #button_eps > 0 then
+    for _,v in ipairs(switch_eps) do
```
Comment (hcarter-775) [inline drivers/SmartThings/matter-switch/src/init.lua:278]: these v's should also be more clear

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 7]
```diff
@@ -144,92 +221,200 @@ local function assign_child_profile(device, child_ep)
...
+    -- Since we do not support bindings at the moment, we only want to count clusters
+    -- that have been implemented as server. This can be removed when we have
+    -- support for bindings.
+    local num_switch_server_eps = 0
+    local main_endpoint = find_default_endpoint(device)
+
+    for _, ep in ipairs(all_eps) do
+      if device:supports_server_cluster(clusters.Switch.ID, ep) then
+        -- Configure MCD for button endpoints
+        if tbl_contains(STATIC_PROFILE_SUPPORTED, #button_eps) then
```
Comment (hcarter-775) [inline drivers/SmartThings/matter-switch/src/init.lua:294]: Can this be a clearer name? From just reading it I had no idea what it did. Sorry for all the naming convention stuff, I know you just ported a lot of this over

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 8]
```diff
@@ -144,92 +221,200 @@ local function assign_child_profile(device, child_ep)
...
         end
-        break
+
+        if device_type_profile_map[id] ~= nil then
+          device:try_update_metadata({profile = device_type_profile_map[id]})
+        end
       end
     end
+  end
+  if device:get_field(IS_PARENT_CHILD_DEVICE) == true then
```
Comment (hcarter-775) [inline drivers/SmartThings/matter-switch/src/init.lua:383]: theoretically, this could be in the above if statement, since only in that statement can this be set to true. 

and nit: remove the " == true "

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 9]
```diff
@@ -144,92 +221,200 @@ local function assign_child_profile(device, child_ep)
...
+  local all_eps = {}
+
+  local new_profile = nil
+
+  local component_map = {}
+  local current_component_number = 2
+  local component_map_used = false
+  local parent_child_device = false
+
+  if #switch_eps > 0 or #button_eps > 0 then
```
Comment (hcarter-775) [inline drivers/SmartThings/matter-switch/src/init.lua:277]:
```suggestion
  if #switch_eps == 0 and #button_eps == 0 then
      return
 end
```
This could remove a level of nesting, and would make the control flow more clear.

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 10]
```diff
@@ -144,92 +221,200 @@ local function assign_child_profile(device, child_ep)
...
+      -- If the device is a parent child device, then set the find_child function on init.
+      -- This is persisted because initialize switch is only run once, but find_child function should be set
+      -- on each driver init.
+      device:set_field(IS_PARENT_CHILD_DEVICE, true, {persist = true})
+    end
+
+    device:set_field(SWITCH_INITIALIZED, true)
+
+    if #button_eps > 0 then
+      local battery_support = false
```
Comment (hcarter-775) [inline drivers/SmartThings/matter-switch/src/init.lua:335]: nit: is_battery_supported instead of battert_support

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 11]
```diff
@@ -144,92 +221,200 @@ local function assign_child_profile(device, child_ep)
...
-            profile = child_profile,
-            parent_device_id = device.id,
-            parent_assigned_child_key = string.format("%d", ep),
-            vendor_provided_label = name
-          }
-        )
+  local button_eps = device:get_endpoints(clusters.Switch.ID, {feature_bitmap=clusters.Switch.types.Feature.MOMENTARY_SWITCH})
+  local all_eps = {}
+
+  local new_profile = nil
```
Comment (hcarter-775) [inline drivers/SmartThings/matter-switch/src/init.lua:270]: nit: profile_name instead of new_profile. 
This is how the naming convention is implemented in matter thermostat and I think it is more accurate

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 12]
```diff
@@ -527,9 +701,91 @@ local function occupancy_attr_handler(driver, device, ib, response)
   device:emit_event(ib.data.value == 0x01 and capabilities.motionSensor.motion.active() or capabilities.motionSensor.motion.inactive())
 end
 
+local function initial_press_event_handler(driver, device, ib, response)
+  if get_field_for_endpoint(device, SUPPORTS_MULTI_PRESS, ib.endpoint_id) then
+    -- Receipt of an InitialPress event means we do not want to ignore the next MultiPressComplete event
+    -- or else we would potentially not create the expected button capability event
+    set_field_for_endpoint(device, IGNORE_NEXT_MPC, ib.endpoint_id, nil)
+  else
```
Comment (hcarter-775) [inline drivers/SmartThings/matter-switch/src/init.lua:709]: why is this not an elseif ?

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 13]
```diff
@@ -527,9 +701,91 @@ local function occupancy_attr_handler(driver, device, ib, response)
...
+    else
+      device:emit_event_for_endpoint(ib.endpoint_id, capabilities.button.button.pushed({state_change = true}))
+    end
+  end
+end
+
+local function multi_press_complete_event_handler(driver, device, ib, response)
+  -- in the case of multiple button presses
+  -- emit number of times, multiple presses have been completed
+  if ib.data and get_field_for_endpoint(device, IGNORE_NEXT_MPC, ib.endpoint_id) ~= true then
```
Comment (hcarter-775) [inline drivers/SmartThings/matter-switch/src/init.lua:742]:
```suggestion
  if ib.data and get_field_for_endpoint(device, IGNORE_NEXT_MPC, ib.endpoint_id) == false then
```
would this not be more clear what the function is? 
Or simply with
```suggestion
  if ib.data and not get_field_for_endpoint(device, IGNORE_NEXT_MPC, ib.endpoint_id) then
```

Comment (hcarter-775) [reply drivers/SmartThings/matter-switch/src/init.lua:742]: Or, and I believe this would be even clearer, we could do:
```suggestion
  if not ib.data or get_field_for_endpoint(device, IGNORE_NEXT_MPC, ib.endpoint_id) then
     log.info_with({hub_logs=true}, "No press received, or press ignored")
     return
  end
```
and then continue the function. This adds better clarity on what is occurring and avoids unclear nesting

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 14]
```diff
@@ -527,9 +701,91 @@ local function occupancy_attr_handler(driver, device, ib, response)
...
+    local press_value = ib.data.elements.total_number_of_presses_counted.value
+    --capability only supports up to 6 presses
+    if press_value < 7 then
```
Comment (hcarter-775) [inline drivers/SmartThings/matter-switch/src/init.lua:745]: similar to above, I suggest doing:
```suggestion
    if press_value >= 7 then
        log.warn("more than 7 button presses. aborting")
        return
   end
```
and then continue with the function

Comment (hcarter-775) [reply drivers/SmartThings/matter-switch/src/init.lua:745]: Or rather, have this be the log:
 log.info(string.format("Number of presses (%d) not supported by capability", press_value))

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 15]
```diff
@@ -144,92 +221,200 @@ local function assign_child_profile(device, child_ep)
...
+      -- This is persisted because initialize switch is only run once, but find_child function should be set
+      -- on each driver init.
+      device:set_field(IS_PARENT_CHILD_DEVICE, true, {persist = true})
+    end
+
+    device:set_field(SWITCH_INITIALIZED, true)
+
+    if #button_eps > 0 then
+      local battery_support = false
+      if device.manufacturer_info.vendor_id ~= HUE_MANUFACTURER_ID and
```
Comment (ctowns) [inline drivers/SmartThings/matter-switch/src/init.lua:336]: Let's create a ticket to update this to use `batteryLevel` instead of just removing the battery capability

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 16]
```diff
@@ -243,6 +428,8 @@ end
 
 local function device_init(driver, device)
   if device.network_type == device_lib.NETWORK_TYPE_MATTER then
+    device:set_component_to_endpoint_fn(component_to_endpoint)
```
Comment (ctowns) [inline drivers/SmartThings/matter-switch/src/init.lua:431]: Can you remind me why these functions needed to be moved to the top of device_init? Just want to make sure I understand why this change was needed

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 17]
```diff
@@ -144,92 +221,200 @@ local function assign_child_profile(device, child_ep)
...
+      table.insert(all_eps, v)
+    end
+    table.sort(all_eps)
+    -- Since we do not support bindings at the moment, we only want to count clusters
+    -- that have been implemented as server. This can be removed when we have
+    -- support for bindings.
+    local num_switch_server_eps = 0
+    local main_endpoint = find_default_endpoint(device)
+
+    for _, ep in ipairs(all_eps) do
```
Comment (ctowns) [inline drivers/SmartThings/matter-switch/src/init.lua:291]: Since this change does not add support for combination switch-button device types, I don't know if we should be considering `all_eps` when going through the initialization. I understand that the combination functionality will be added in a future change, but currently this solution seems to provide some support for combination devices by combining eps, which could potentially result in unexpected pathways when onboarding a combination device that may result in matching to a profile that doesn't exist. It may be easier to just separate the switch and button ep handling all together such that if switch_eps > 0, then handle making switch child devices. elseif button eps > 0, then make button endpoint to component maps. In this case, switch takes precendent such that if a device supports both switch and button EPs, only the switch ones would be present. But it seems potentially buggy to combine the ep handling in this init function when we do not support combination devices throughout the rest of the driver.

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 13 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua:742]: I think the second option you recommended is the best option, since we need to set `IGNORE_NEXT_MPC` to nil at the end of the function whether this conditional is true or not.

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 14 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua:745]: Similar to a previous comment you suggested, `IGNORE_NEXT_MPC` needs to be set to nil at the end of the function in all cases, so while an early return would be nice I'm not sure it will work. I would prefer to keep this as-is for this reason.

Comment (hcarter-775) [reply drivers/SmartThings/matter-switch/src/init.lua:745]: Alright, good point. Sounds good to me :+1:

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 16 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua:431]: This was a leftover from when the deferred configuration for buttons wasn't in place, and I believe they can safely be restored to their original location.

File: drivers/SmartThings/matter-button/src/init.lua [hunk 18]
```diff
@@ -1,3 +1,10 @@
+-- As of July 2024, the matter-button driver has been deprecated in favor of being integrated into the matter-switch
```
[the remark this answers is not in the document]
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-button/src/init.lua]: Updated in de641ad73df7965367fa355d0b54a3570ccec054

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 1 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua:111]: Updated to `create_multi_press_values_list` in de641ad73df7965367fa355d0b54a3570ccec054

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 2 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua:114]: This is for devices that support multipress (3 presses or greater). I added a comment to make this more clear - let me know if I should add more detail

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 3 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua]: Updated to `ep` to make it clear that we are iterating through the endpoints

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 6 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua]: Updated to `ep` to make it clear that we are iterating through the endpoints

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 7 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua]: Updated to `STATIC_BUTTON_PROFILE_SUPPORTED` to be more clear. This is just making sure that a profile exists for the given amount of button endpoints before attempting to index into the component to endpoint map.

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 8 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua]: After implementing your other suggestion to remove the nesting from the above if statement, I think that that addressed this comment as well. I also removed the `== true`

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 9 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua]: Updated as suggested in de641ad73df7965367fa355d0b54a3570ccec054

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 10 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua]: I would prefer to leave this as-is

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 11 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua]: Updated as suggested

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 12 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua]: Updated to an if/elseif/elseif structure in de641ad73df7965367fa355d0b54a3570ccec054

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 16 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua]: Restored to their original position in de641ad73df7965367fa355d0b54a3570ccec054

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 4 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua:176]: I would prefer to leave this as-is

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 5 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua:269]: I would prefer to leave this as-is for now

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 19]
```diff
@@ -335,7 +513,7 @@ local function handle_set_hue(driver, device, cmd)
     device:send(req)
   else
     log.warn("Device does not support huesat features on its color control cluster")
- end
+  end
```
Comment (hcarter-775) [inline drivers/SmartThings/matter-switch/src/init.lua:516]: nit: remove whitespace

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 20]
```diff
@@ -144,92 +222,202 @@ local function assign_child_profile(device, child_ep)
...
-            vendor_provided_label = name
-          }
+            {
+              type = "EDGE_CHILD",
+              label = name,
+              profile = child_profile,
+              parent_device_id = device.id,
+              parent_assigned_child_key = string.format("%d", ep),
+              vendor_provided_label = name
+            }
```
Comment (hcarter-775) [inline drivers/SmartThings/matter-switch/src/init.lua:321]: nit: remove added whitespace?

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 10 above)
Comment (hcarter-775) [reply drivers/SmartThings/matter-switch/src/init.lua]: sure!

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 20 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua]: Fixed in 56f40cf534af547dee93afcff6d524d82419be86

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 19 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua:509]: Was this in a previous commit? Looks like there isn't extra whitespace anymore

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 17 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua]: Addressed in 56f40cf534af547dee93afcff6d524d82419be86. I separated the initialization of switch and button eps as suggested. This also helped consolidate things a bit.

Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua]: Let me know if the changes line up with what you had in mind!

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 19 above)
Comment (hcarter-775) [reply drivers/SmartThings/matter-switch/src/init.lua:509]: still sees to be there, not sure. It's at the little extra space in front of `end`, not trailing whitespace

Comment (hcarter-775) [reply drivers/SmartThings/matter-switch/src/init.lua:509]: nevermind, sorry about that. everything is as expected here :+1:

Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua:509]: Ok sweet!

**Review by hcarter-775 — APPROVED** [review]
All of my comments have been accounted for, and otherwise this seems like a pretty safe code change. Awaiting Cooper's approval on the points he noted, it lgtm!

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 21]
```diff
@@ -168,68 +301,116 @@ local function initialize_switch(driver, device)
             vendor_provided_label = name
           }
         )
+        current_component_number = current_component_number + 1
```
Comment (ctowns) [inline drivers/SmartThings/matter-switch/src/init.lua:304]: current_component_number shouldn't increase when creating a child device since this PR does not support combo button/switch devices

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 22]
```diff
@@ -168,68 +301,116 @@ local function initialize_switch(driver, device)
...
-    local id = 0
-    for _, ep in ipairs(device.endpoints) do
-      -- main_endpoint only supports server cluster by definition of get_endpoints()
-      if main_endpoint == ep.endpoint_id then
-        for _, dt in ipairs(ep.device_types) do
-          -- no device type that is not in the switch subset should be considered.
-          if (ON_OFF_SWITCH_ID <= dt.device_type_id and dt.device_type_id <= ON_OFF_COLOR_DIMMER_SWITCH_ID) then
-            id = math.max(id, dt.device_type_id)
+
+  if device:get_field(IS_PARENT_CHILD_DEVICE) then
```
[the remark this answers is not in the document]
Comment (ctowns) [reply drivers/SmartThings/matter-switch/src/init.lua:332]: Actually, this is already handled in device_init isn't it? This shouldn't be needed here

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 23]
```diff
@@ -168,68 +301,116 @@ local function initialize_switch(driver, device)
...
+        profile_name = string.format("%d-button-battery", #button_eps)
+      else
+        profile_name = string.format("%d-button", #button_eps)
+      end
+    elseif not battery_support then
+      -- a battery-less button/remote (either single or will use parent/child)
+      profile_name = "button"
+    end
+
+    if profile_name then
```
Comment (ctowns) [inline drivers/SmartThings/matter-switch/src/init.lua:384]: Coudl combine this conditional with the one below

File: drivers/SmartThings/matter-switch/src/init.lua [hunk 24]
```diff
@@ -253,11 +434,11 @@ local function device_init(driver, device)
       -- create child devices as needed for multi-switch devices
       initialize_switch(driver, device)
     end
-    device:set_component_to_endpoint_fn(component_to_endpoint)
-    device:set_endpoint_to_component_fn(endpoint_to_component)
     if device:get_field(IS_PARENT_CHILD_DEVICE) == true then
       device:set_find_child(find_child)
     end
+    device:set_component_to_endpoint_fn(component_to_endpoint)
```
Comment (ctowns) [inline drivers/SmartThings/matter-switch/src/init.lua:440]: nit: can these be put in their original location above the `if device:get_field(IS_PARENT_CHILD_DEVICE) == true then` above to reduce the unneccessary diff?

File: drivers/SmartThings/matter-switch/src/test/test_matter_multi_button.lua [hunk 25]
```diff
@@ -0,0 +1,732 @@
+local test = require "integration_test"
+local capabilities = require "st.capabilities"
+local t_utils = require "integration_test.utils"
+local utils = require "st.utils"
+local dkjson = require "dkjson"
+
+local clusters = require "st.matter.generated.zap_clusters"
+local button_attr = capabilities.button.button
+
+-- Mock a 4-button device using endpoints non-consecutive endpoints
```
Comment (ctowns) [inline drivers/SmartThings/matter-switch/src/test/test_matter_multi_button.lua:10]: nit: a mock 5-button device

**Review by ctowns — COMMENTED** [review]
I left a few more small comments but this is my final review and I think it's ready. Has this also been regression tested with some of the switch/button devices on the VDA? For example, confirming a light, button, and composed bridged devices work would be a good test.

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 21 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua]: Fixed in 340df0f5c6c6d95d4d9b7c19e2b70b4b2fb646e2

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 22 above)
[the remark this answers is not in the document]
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua]: Removed in 340df0f5c6c6d95d4d9b7c19e2b70b4b2fb646e2

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 23 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua:379]: Updated in 340df0f5c6c6d95d4d9b7c19e2b70b4b2fb646e2

File: drivers/SmartThings/matter-switch/src/init.lua (hunk 24 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/init.lua]: Put back in original position in 340df0f5c6c6d95d4d9b7c19e2b70b4b2fb646e2

File: drivers/SmartThings/matter-switch/src/test/test_matter_multi_button.lua (hunk 25 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-switch/src/test/test_matter_multi_button.lua]: Fixed in 340df0f5c6c6d95d4d9b7c19e2b70b4b2fb646e2

**Comment by nickolas-deboom** [comment]
> I left a few more small comments but this is my final review and I think it's ready. Has this also been regression tested with some of the switch/button devices on the VDA? For example, confirming a light, button, and composed bridged devices work would be a good test.

I tested several virtual switch devices as well as a composed bridge device with the VDA and everything is functioning as expected. The Generic Switch virtual device does not onboard properly so I tested the button functionality with an IKEA remote, which functioned as expected as well.
